### PR TITLE
Userguide section and whitepaper on user change management.

### PR DESCRIPTION
**Pull request by pp-mo**
Suggestions for clarifying change management from the user POV
 = releases + versions, deprecations, iris.FUTURE

This is mostly for discussion at this point.
It is now pretty much complete + correct to my original conception.
However, @rhattersley already suggested that this information might with advantage be a bit "generalised" and then "pushed back" onto the SemVer project https://github.com/mojombo/semver.
As I was most of the way toward providing this a standalone Iris docs, I've just carried on with that for this first draft. 

This issue is mentioned as a possible for 1.10, #1948 
which would be nice as we are targetting a lot of deprecations in 1.10.
While may be relevant there, it certainly isn't essential.

### CODE REVIEW

File: docs/iris/src/userguide/change_management_goals.txt [hunk 1]
```diff
@@ -0,0 +1,9 @@
+To reduce code maintenance problems to an absolute minimum, the Iris project defines
+carefully defined change management procedures to ensure that :
```
Comment (QuLogic) [inline docs/iris/src/userguide/change_management_goals.txt]: defines carefully defined? A bit repetitive.

File: docs/iris/src/whitepapers/change_management.rst [hunk 2]
```diff
@@ -0,0 +1,393 @@
...
+=============================================================
+
+Checklist :
+
+* when a new **testing or candidate version** is announced
+    if convenient, test your working legacy code against it and report any problems.
+
+* when a new **minor version is released**
+
+    * review the 'Whatsnew' documentation to see if it introduces any
```
Comment (QuLogic) [inline docs/iris/src/whitepapers/change_management.rst]: No need to have 'Whatsnew' as one word, since it's not a link to anything that might be named that way.

File: docs/iris/src/whitepapers/change_management.rst [hunk 3]
```diff
@@ -0,0 +1,393 @@
...
+.. _iris_backward_compatibility:
+
+Backwards compatibility
+-----------------------
+
+"Backwards-compatible" changes are those that leave all possible existing API
+usages unchanged (see :ref:`terminology <iris_api>` below).
+Only such changes may be included in minor releases.
+
+the following are examples of backward-compatible changes :
```
Comment (QuLogic) [inline docs/iris/src/whitepapers/change_management.rst]: Capital T.

File: docs/iris/src/whitepapers/change_management.rst [hunk 4]
```diff
@@ -0,0 +1,393 @@
...
+
+    * changes to documentation
+    * adding to a module : new submodules, functions, classes or properties
+    * adding to a class : new methods or properties
+    * adding to a function or method : new optional arguments or keywords
+
+The following are examples of **non-** backward-compatible changes :
+
+    * removing a call
```
Comment (QuLogic) [inline docs/iris/src/whitepapers/change_management.rst]: Should this be "removing a function"? I don't see how removing a call is necessarily backwards-incompatible, unless it's to some user-defined callback.

File: docs/iris/src/whitepapers/change_management.rst [hunk 5]
```diff
@@ -0,0 +1,393 @@
...
+
+
+.. _iris_api:
+
+Terminology : API, features, usages and behaviours
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The API is the components of the iris module and submodules which are
+"public" :  In Python, by convention, this means everything whose name does not
+have a leading underscore "_".
```
Comment (QuLogic) [inline docs/iris/src/whitepapers/change_management.rst]: A _single_ leading underscore, because something like `__init__` is definitely part of the API.

File: docs/iris/src/whitepapers/change_management.rst [hunk 6]
```diff
@@ -0,0 +1,393 @@
...
+This means all public modules and submodules, and their contained classes,
+public data and properties, functions and methods.
+
+A "feature" of the API includes public objects as above, but may also be used
+more loosely to indicate a class or mode of behaviour, for example when a
+keyword has a specific value, like "interpolate(mode='linear')".
+
+A "usage" is any code referring to public API elements, for example :
+
+        * `print iris.thing`
```
Comment (QuLogic) [inline docs/iris/src/whitepapers/change_management.rst]: Please use the `print` function, not statement.

File: docs/iris/src/whitepapers/change_management.rst [hunk 7]
```diff
@@ -0,0 +1,393 @@
...
+
+    * "<major>.<minor>" denote the software release version.
+
+    * A non-zero "<micro>" denotes a bugfix version, thus a release "X.Y.0" may
+      be followed by "X.Y.1", "X.Y.2" etc, which *only* differ by containing
+      bugfixes.  Any bugfix release supercedes its predecessors, and does not
+      change any (valid) APIs or behaviour :  there should be no reason not to
+      replace a given version with its latest bugfix successor.
+
+    * "<extension>" is blank for formal releases.  It used to indicate
```
Comment (QuLogic) [inline docs/iris/src/whitepapers/change_management.rst]: This section might need some tweaks to follow PEP440, but that might be a discussion for a different place. I don't think what's defined here would pass PEP440, though it might only result in a warning from `pip` at this point.

File: docs/iris/src/whitepapers/change_management.rst [hunk 8]
```diff
@@ -0,0 +1,393 @@
...
+      provisional software for testing :  The version of general Iris code in
+      development is labelled "-DEV", and release candidates for testing during
+      the release process are labelled "-rc1", "-rc2" etc.
+      For development code, the version number is that of the *next* release,
+      which this code version is progressing towards, e.g. "1.2-DEV" for code
+      following the 1.1 release and eventually giving rise to "1.2".
+
+.. note::
+    Our use of "-<extension>" is typical, but does not follow strict SemVer
+    principles.
```
Comment (QuLogic) [inline docs/iris/src/whitepapers/change_management.rst]: If that's the case, maybe it should be tweaked a little to match PEP440.

Comment (rhattersley) [inline docs/iris/src/whitepapers/change_management.rst]: :+1: for standards :grin:

File: docs/iris/src/whitepapers/change_management.rst [hunk 9]
```diff
@@ -0,0 +1,393 @@
...
+.. _iris_api:
+
+Terminology : API, features, usages and behaviours
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The API is the components of the iris module and submodules which are
+"public" :  In Python, by convention, this means everything whose name does not
+have a leading underscore "_".
+This means all public modules and submodules, and their contained classes,
+public data and properties, functions and methods.
```
Comment (rhattersley) [inline docs/iris/src/whitepapers/change_management.rst]: Not quite all, as clarified in our [deprecation page](http://scitools.org.uk/iris/docs/latest/developers_guide/deprecations.html#public-api):

> with the exception of the `iris.experimental` and `iris.tests` packages.

But perhaps we need to tighten our definition further? Perhaps by restricting to the contents of `__all__`? For example, if I import iris and look for names without a single leading underscore but that _aren't_ in `__all__` I get:

['absolute_import', 'analysis', 'aux_factory', 'config', 'contextlib', 'coord_systems', 'coords', 'cube', 'division', 'exceptions', 'experimental', 'fileformats', 'filter', 'input', 'io', 'iris', 'itertools', 'logging', 'map', 'os', 'print_function', 'range', 'six', 'std_names', 'threading', 'time', 'util', 'zip']

That's a mix of std lib imports, `__future__` features, iris sub-packages/modules, and `six` imports. Surely we don't consider them part of our public API?

File: docs/iris/src/whitepapers/change_management.rst (hunk 8 above)
Comment (pp-mo) [inline docs/iris/src/whitepapers/change_management.rst]: Our current incompatibility with SemVer is limited to defining versions such as "1.9.0rc1",
instead of "1.0.-rc.1" : adding a hyphen and a dot.
See the tags ...
We did at one point have "1.7.1-rc.1" thru "1.7.3-rc.1", but then reverted to "1.7.4-rc1".
Since then, we have dropped both the hyphen and the dot.
The current Developer guide also specifically requires this form, which is **PEP440 but not SemVer**.

The big problem is that PEP440 and SemVer are **mutually incompatible** in this respect :frowning: 

@SemVer 
> "A pre-release version MAY be denoted by appending a hyphen and a series of dot separated identifiers immediately following the patch version"
  
@PEP440 
> "[N!]N(.N)*[{a|b|rc}N][.postN][.devN]"

**IMHO** we should just change this + have done with it.

Comment (pp-mo) [inline docs/iris/src/whitepapers/change_management.rst]: **IMHO**
we should standardise on SemVer + have done with it, if at all possible.

The only other question I have about conformance is whether our "FUTURE" strategy truly complies with SemVer ...
I wonder if it's a problem that you can change the behaviour of a piece of code by an unrelated assignment elsewhere -- though I suppose in Python you can always do that by assigning to method properties etc.

File: docs/iris/src/whitepapers/change_management.rst (hunk 9 above)
Comment (pp-mo) [inline docs/iris/src/whitepapers/change_management.rst]: I think this is an issue illustrated by https://github.com/SciTools/iris/pull/1962

That removed various public-named items from the iris.fileformats.grid module.
These, by my reckoning :
```CENTRE_TITLES, TIME_RANGE_INDICATORS, PROCESSING_TYPES, TIME_CODES_EDITION1, TIME_CODES_EDITION2, unknown_string, GribDataProxy```

Maybe that is ok, if they were excluded from the "__all__" in the original ?
It means their docstrings aren't published, and they are omitted from a "from iris.fileformats.grib import *"

Is that good enough to consider them ***private*** ??

**Comment by pp-mo** [comment]
>  if I import iris and look for names without a single leading underscore ... iris sub-packages ...

It's surely ok to exclude local imports, though it certainly is an annoying glitch and a backwards incompatibility if code fails because we removed an automatic submodule import
 
E.G. what if importing ```iris``` did *not* result in importing ```iris.cube``` ?

I think @rhattersley already suggested that all internal imports could be given private internal names, to avoid this kind of problem.
Changing that now will doubtless upset lots of existing user code, but it we want it we should get on right now and introduce a deprecation for it in 1.10

I've put a note in the 1.10 ticket #1948 for now
However, I'm not sure if this is actually  feasible as it stands ...
```
>>> import iris
>>> hasattr(iris, 'proxy')
False
>>> from iris import proxy as _proxy
/home/h05/itpp/git/iris/iris_main/lib/iris/proxy.py:34: UserWarning: iris.proxy is deprecated in Iris v1.9. Please use lazy imports instead.
  warnings.warn('iris.proxy is deprecated in Iris v1.9. Please use lazy '
>>> hasattr(iris, 'proxy')
True
>>> 
```

File: docs/iris/src/whitepapers/change_management.rst (hunk 9 above)
Comment (rhattersley) [inline docs/iris/src/whitepapers/change_management.rst]: For reference, the [Python docs](https://docs.python.org/2/reference/simple_stmts.html?highlight=__all__#the-import-statement) have a description of `__all__` that includes:

> It is intended to avoid accidentally exporting items that are not part of the API (such as library modules which were imported and used within the module).

So I'm not alone in suggesting this pattern.

But they also state:

> If `__all__` is not defined, the set of public names includes all names found in the module’s namespace which do not begin with an underscore character (`'_'`).

So we'd need to start being a bit more thorough with our `__all__` definitions.

**Comment by rhattersley** [comment]
> I think @rhattersley already suggested that all internal imports could be given private internal names, to avoid this kind of problem.

I'm now :-1: on this. I mentioned the idea yesterday offline, but it doesn't address the whole problem. Instead I think it'd be better to use the existing `__all__` convention.

File: docs/iris/src/whitepapers/change_management.rst (hunk 9 above)
Comment (rhattersley) [inline docs/iris/src/whitepapers/change_management.rst]: NB. Currently the Jupyter notebook doesn't respect `__all__`, but does ignore leading underscores. I've raised https://github.com/ipython/ipykernel/issues/129 to see if that can be improved.

File: docs/iris/src/whitepapers/change_management.rst (hunk 8 above)
Comment (rhattersley) [inline docs/iris/src/whitepapers/change_management.rst]: > we should standardise on SemVer + have done with it, if at all possible.

I'm leaning towards sticking with [PEP 440](https://www.python.org/dev/peps/pep-0440/) as the way of encoding well-ordered versions (as that is what setuptools, etc. should target, and we are talking about a Python package after all) but keeping the semantic aspects from [SemVer](http://semver.org/).

File: docs/iris/src/whitepapers/change_management.rst (hunk 9 above)
Comment (rhattersley) [inline docs/iris/src/whitepapers/change_management.rst]: > I think this is an issue illustrated by #1962

I think you meant #1755.

Comment (pp-mo) [inline docs/iris/src/whitepapers/change_management.rst]: > I think you meant #1755.

Yes, probably !

**Comment by pp-mo** [comment]
Thanks @rhattersley 
This is a jolly useful discussion (for me, anyway)

> But they also state:
> >  If ```__all__``` is not defined, the set of public names includes all names found in the module’s namespace which do not begin with an underscore character ```('_')```.
> 
>  So we'd need to start being a bit more thorough with our ```__all__``` definitions.

All true, but for now all that really matters is what we include in **our** definition of "public"
 \- and I now think this *is* a good candidate to include.
Given that, I think I should enhance the formal definition in the Developer's Guide, and just reference  that here rather than repeating all the gory details :  Applies also to the point you already made about ```experimental```.

**Comment by pp-mo** [comment]
> However, I'm not sure if this is actually feasible as it stands ...
> >
```
>>> import iris
>>> hasattr(iris, 'proxy')
>>> from iris import proxy as _proxy
```

To be clear, this example really isn't quite the point...
Though you can't avoid (e.g.) iris having a 'proxy' property if it has ever been imported,
the *real* problem is more like this one :
```
>>> import iris.cube
>>> iris.cube.datetime.MIN_YEAR
1
```
I.E. 'datetime' appears as a public property of iris.cube.
That we can avoid.

**Comment by pp-mo** [comment]
Note: I'm going to let this stew a bit longer before doing any more, in case there are further comments.

**Comment by pp-mo** [comment]
@QuLogic thanks, not much to say : all your comments were true !

**Comment by pp-mo** [comment]
@rhattersley suggested I could consider pushing some of this content back into the SemVer document.
So I thought about what here is SemVer and what not.

My conclusions are...

The SemVer provisions:
 * backwards-incompatible changes can only occur at major releases
 * deprecations can only be introduced at a minor release (not bugfix)
 * API additions and changes can only be introduced at a minor release (not bugfix)
 * API means "public"
 * there are no definitions of certain terms used :
  * API
  * public
  * behavior / functionality
  * deprecation (what "marking API functionality as deprecated" exactly means)

Additional provisions from Iris (as I'm attempting to rationalise them):
 * key certainties we are adding:
   * **only** deprecated features may be changed at a major release
   * deprecated functionality will **always** emit runtime deprecation warnings
     * ( for user code that 'could' invoke deprecated functionality, e.g. with different data )
   * ("therefore":) code that ran in the version preceding a major release, without deprecation warnings, will continue to function unchanged
 * deprecations
   * will always involve documentation
   * will always raise runtime warnings for code where it 'could' make a difference
     * ( though it still isn't clear to me whether we can *always* deliver this )
 * iris.FUTURE : central control for compatibility-breaking modal changes
 * terms we are trying to define more precisely :
   * what public API means
   * 'backwards compatible' means any conceivable existing code should behave as before
   * what 'behaviour' means
     * ( return values and side effects, but somewhat limited : excluding warnings and printed output, possibly ? )

File: docs/iris/src/whitepapers/change_management.rst [hunk 10]
```diff
@@ -0,0 +1,393 @@
...
+
+
+Key principles you can rely on
+==============================
+
+Iris code editions are published as defined version releases, with a given
+major and minor version number in the version name, "major.minor.xxx",
+as explained in the :ref:`releases section <iris_change_releases>` below.
+
+    * Code that currently works should **still work**, and behave exactly the
```
Comment (pelson) [inline docs/iris/src/whitepapers/change_management.rst]: This is a fallacy. It is not possible to change *anything* without breaking something else. https://xkcd.com/1172/

**Comment by pp-mo** [comment]
> This is a fallacy. It is not possible to change anything without breaking something else. https://xkcd.com/1172/

Classik ! :grinning: 

But Seriously..
I had a good chat with @rhattersley yesterday about what it would mean to separate iris.fileformats.grib off into a separate package/project, and that provided an interesting example.

This makes very clear a point which I had not properly considered :  That any "behaviour" as I've tried to define it here (aka all outcomes of some code) potentially changes with all dependencies, including upgrades non.

It means :
  * anything we claim to be "exactly the same" only applies to changing (upgrading) Iris *and nothing else* (unless explicit dependencies changed.

**Comment by pp-mo** [comment]
> This is a fallacy. It is not possible to change anything without breaking something else. https://xkcd.com/1172/

Classik ! :grinning: 

But Seriously..
I had a good chat with @rhattersley yesterday about what it would mean to separate iris.fileformats.grib off into a separate package/project, as you suggested :  so that we can release Iris 1.10 *without* needing to first resolve all the ongoing efforts to deprecate parts of it and provide alternatives for everything.
> ( The idea would be to spinoff "iris_grib", initially equivalent to the current "iris.fileformats.grib", but then evolving seperately.
In particular, we can then deprecate and remove the "old" loader without requiring an Iris major version release. )

This example makes clear a point which I had not properly considered in the proposals here :
Any "behaviour" (as here, i.e. all outcomes of any code) can potentially change when any dependency is changed (e.g. upgraded), to the point of breaking (e.g. dependency major version change).

That means we *can* only completely fossilize behaviour by pinning the version of all dependencies (that is, at least their *major* versions, where the concept applies).
Which seems pretty crazy and meaningless. 
Instead, I guess the correct approach is to **redefine "behaviour" in terms of the calls on component dependencies, not total results** ??
Any ideas how best to state this ?

I think we should still accept that if a version upgrade in a dependency actually *breaks* something, then that is a bug and we need to limit that version until we fix it.
This is currently true for matplotlib and numpy : we aren't fully compatible with latest versions.

**Comment by pp-mo** [comment]
@marqh what are your thoughts on this ?
I'd like us to firm up on it, but this may be altogether too wordy

**Comment by pp-mo** [comment]
Latest revisions...
 * I have attempted to respond to all the simple improvement suggestions already posted
 * I have attempted to define more precisely what we mean by "public"
 * I have attempted to define more vaguely (!) what we mean by "same behaviour"
 * I haven't yet taken any position on the PEP440 / SemVer version string formatting issue.

**Comment by pp-mo** [comment]
Other issues mentioned in discussion above, but still not addressed in the text :
 * imported modules-within-modules (as excluded from "public" API)
 * segregated description of SemVer principles, and where we add to them

**Comment by marqh** [comment]
ping @marqh

File: docs/iris/src/userguide/change_management_goals.txt [hunk 11]
```diff
@@ -0,0 +1,9 @@
+To reduce code maintenance problems to an absolute minimum, Iris applies
```
Comment (marqh) [inline docs/iris/src/userguide/change_management_goals.txt]: i recommend removing these sections from the user guide

this section is already in use in the white paper, so the content is preserved.

File: docs/iris/src/userguide/code_maintenance.rst [hunk 12]
```diff
@@ -0,0 +1,38 @@
+Code Maintenance
```
Comment (marqh) [inline docs/iris/src/userguide/code_maintenance.rst]: I think this is covered in the white paper.  This section could be removed, to try and keep the user guide more tightly focussed

**Comment by marqh** [comment]
i think it is worth considering keeping this to the white paper, and not adding a section to the user guide

how much benefit do we get from the extra user guide section as well?

**Comment by pp-mo** [comment]
> not adding a section to the user guide

I agree the userguide is rather bloated now.
My concern is that this topic should definitely be on a list of "things the user should know, that they may not know they need to know".
Can you suggest a better approach that satisfies that need ?

The userguide contents listing was just the **_only_** existing place I could think of that performs that function.  Obviously, though, it does include a good few chapters that are firmly in the "specialist interest : come back later" category.
Right now, we don't have a clear 'read this first' section at present, and perhaps we should...

**Comment by marqh** [comment]
ok, I'm sold; let's accept this now.  Any reworking of the docs can include a review of


> Right now, we don't have a clear 'read this first' section at present, and perhaps we should...
